File: leetcode/1043-grid-illumination/solution.go
func gridIllumination(n int, lamps [][]int, queries [][]int) []int {
    lampByXY := make(map[[2]int]struct{})
    lampByRow := make(map[int]int)
    lampByCol := make(map[int]int)
    lampByDeltaXY := make(map[int]int)
    lampByDeltaMXY := make(map[int]int)

    for _, l := range lamps {
        lx, ly := l[0], l[1]
        pos := [2]int{lx, ly}
        if _, exists := lampByXY[pos]; !exists {
            lampByXY[pos] = struct{}{}
            lampByRow[lx]++
            lampByCol[ly]++
            lampByDeltaXY[lx-ly]++
            lampByDeltaMXY[-lx-ly]++
        }
    }

    ret := make([]int, len(queries))

    for qi, q := range queries {
        qx, qy := q[0], q[1]
        lit := lampByRow[qx] > 0 ||
            lampByCol[qy] > 0 ||
            lampByDeltaXY[qx-qy] > 0 ||
            lampByDeltaMXY[-qx-qy] > 0
        if lit {
            ret[qi] = 1
        } else {
            ret[qi] = 0
        }

        for x := qx-1; x <= qx+1; x++ {
            for y := qy-1; y <= qy+1; y++ {
                pos := [2]int{x, y}
                if _, exists := lampByXY[pos]; exists {
                    delete(lampByXY, pos)
                    lampByRow[x]--
                    lampByCol[y]--
                    lampByDeltaXY[x-y]--
                    lampByDeltaMXY[-x-y]--
                }
            }
        }
    }

    return ret
}

File: leetcode/1043-grid-illumination/solution.rs
impl Solution {
    pub fn grid_illumination(n: i32, lamps: Vec<Vec<i32>>, queries: Vec<Vec<i32>>) -> Vec<i32> {
        use std::collections::{HashMap, HashSet};
        let mut lamp_by_xy: HashSet<(i32, i32)> = HashSet::new();
        let mut lamp_by_row: HashMap<i32, i32> = HashMap::new();
        let mut lamp_by_col: HashMap<i32, i32> = HashMap::new();
        let mut lamp_by_delta_xy: HashMap<i32, i32> = HashMap::new();
        let mut lamp_by_delta_mxy: HashMap<i32, i32> = HashMap::new();
        let mut ret = Vec::new();
        for l in lamps {
            let lx = l[0];
            let ly = l[1];
            if lamp_by_xy.insert((lx, ly)) {
                *lamp_by_row.entry(lx).or_default() += 1;
                *lamp_by_col.entry(ly).or_default() += 1;
                *lamp_by_delta_xy.entry(lx-ly).or_default() += 1;
                *lamp_by_delta_mxy.entry(-lx-ly).or_default() += 1;
            }
        }
        for q in queries {
            let qx = q[0];
            let qy = q[1];
            let lit = *lamp_by_row.entry(qx).or_default() > 0 ||
                *lamp_by_col.entry(qy).or_default() > 0 ||
                *lamp_by_delta_xy.entry(qx-qy).or_default() > 0 ||
                *lamp_by_delta_mxy.entry(-qx-qy).or_default() > 0;
            ret.push(if lit {1} else {0});
            for x in qx-1..=qx+1 {
                for y in qy-1..=qy+1 {
                    if lamp_by_xy.remove(&(x,y)) {
                        *lamp_by_row.entry(x).or_default() -= 1;
                        *lamp_by_col.entry(y).or_default() -= 1;
                        *lamp_by_delta_xy.entry(x-y).or_default() -= 1;
                        *lamp_by_delta_mxy.entry(-x-y).or_default() -= 1;
                    }
                }
            }
        }
        return ret;
    }
}
